rfc/rfc-95/rfc-95.md

```diff
@@ -0,0 +1,263 @@
+    return scanContext.isStreaming() ? Boundedness.CONTINUOUS_UNBOUNDED : Boundedness.BOUNDED;
+  }
+
+  @Override
+  public SourceReader<T, HudiSourceSplit> createReader(SourceReaderContext readerContext) {
+    HudiSourceReaderMetrics metrics =
+        new HudiSourceReaderMetrics(readerContext.metricGroup(), tableName);
+    return new HudiSourceReader<>(
+        emitter, metrics, readerFunction, splitComparator, readerContext);
+  }
+
+  @Override
+  public SplitEnumerator<HudiSourceSplit, HudiEnumeratorState> createEnumerator(
+      SplitEnumeratorContext<HudiSourceSplit> enumContext) {
+    return createEnumerator(enumContext, null);
+  }
+
+  @Override
+  public SplitEnumerator<HudiSourceSplit, HudiEnumeratorState> restoreEnumerator(
+      SplitEnumeratorContext<HudiSourceSplit> enumContext, HudiEnumeratorState enumState) {
+    return createEnumerator(enumContext, enumState);
+  }
+
+  @Override
+  public SimpleVersionedSerializer<HudiSourceSplit> getSplitSerializer() {
+    return new HudiSourceSplitSerializer(scanContext.caseSensitive());
+  }
+
+  @Override
+  public SimpleVersionedSerializer<HudiEnumeratorState> getEnumeratorCheckpointSerializer() {
+    return new HudiEnumeratorStateSerializer(scanContext.caseSensitive());
+  }
+  
+  // ...
+}
+```
+
+### Hudi Split
+Similar to existing MergeOnReadInputSplit that implements InputSplit API, A new MergeOnReadSourceSplit and CdcSourceSplit will be implemented with the new SourceSplit interface.
+FlinkRowDataReaderContext will be extended to be able to apply HudiSourceSplit and return a CloseableIterator that iterates RecordsWithSplitIds. Base on the change, the HudiSourceSplitReader
+will be skeleton as below.
+
+```java
+
+class HudiSourceSplitReader implements SplitReader<RecordAndPosition<RowData>, HudiSourceSplit> {
+  private final FlinkRowDataReaderContext readerContext;
+  private final SerializableComparator<HudiSourceSplit> splitComparator;
+  private final int indexOfSubtask;
+  private final Queue<HudiSourceSplit> splits;
+
+  private CloseableIterator<RecordsWithSplitIds<RecordAndPosition<RowData>>> currentReader;
+  private HudiSourceSplit currentSplit;
+  private String currentSplitId;
+
+  HudiSourceSplitReader(
+      FlinkRowDataReaderContext readerContext,
+      SerializableComparator<HudiSourceSplit> splitComparator,
+      SourceReaderContext context) {
+    this.readerContext = readerContext;
+    this.splitComparator = splitComparator;
+    this.indexOfSubtask = context.getIndexOfSubtask();
+    this.splits = Queues.newArrayDeque();
+  }
+  
+  @Override
+  public RecordsWithSplitIds<RecordAndPosition<T>> fetch() throws IOException {
+    metrics.incrementSplitReaderFetchCalls(1);
+    if (currentReader == null) {
+      HudiSourceSplit nextSplit = splits.poll();
+      if (nextSplit != null) {
+        currentSplit = nextSplit;
+        currentSplitId = nextSplit.splitId();
+        currentReader = readerContext.apply(currentSplit);
```

Comment:
is the reader a file group reader? We don't really have `HoodieReaderContext.apply` BTW.